nexus/src/app/sled.rs

```diff
@@ -0,0 +1,145 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Sleds, and the hardware and services within them.
+
+use crate::context::OpContext;
+use crate::db;
+use crate::db::identity::Asset;
+use crate::db::lookup::LookupPath;
+use crate::db::model::DatasetKind;
+use crate::internal_api::params::ZpoolPutRequest;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use sled_agent_client::Client as SledAgentClient;
+use std::net::{SocketAddr, SocketAddrV6};
+use std::sync::Arc;
+use uuid::Uuid;
+
+impl super::Nexus {
+    // Sleds
+
+    // TODO-robustness we should have a limit on how many sled agents there can
+    // be (for graceful degradation at large scale).
+    pub async fn upsert_sled(
```

Comment:
I gather we've grouped things by the noun that they seem to relate to (e.g., "sled" here).  I find it a little weird to combine the internal vs. external functions in this way.  That is, `upsert_sled` is eventually going to do a whole bunch of work in response to the new information that a Sled Agent has started up...that all feels pretty unrelated to the public-facing sled CRUD functions.

sled_allocate() is different than both of these.

I felt a similar dissonance around saga.rs, where `execute_saga()` really has little to do with the saga fetch/list functions.

I don't know what else to suggest, though.  Maybe a better structure will emerge when we flesh these things out more.  (e.g., maybe `upsert_sled` will eventually become part of a `stationkeeping` submodule that monitors the rack's health and takes appropriate action as data comes in).

Reply:
I'm still a little split on this - I'm not sure if this layer even needs to make a distinction of "internal" vs "external" APIs. After all, the `internal_api` and `external_api` directories in Nexus are sorta making this distinction for us, in a literal sense.

After all, if both interfaces (external and internal) poke at a sled, what's the issue with grouping them together? They'll be accessing the same underlying structures, and the implementations may end up needing to be aware of each other. As an example: The "notify_instance_changed" implementation is internal (called only by sled agent) but the external "change instance state" methods are aware of the sled agent "owning" instance state data.

Definitely agree though, we can keep iterating on this - this is my first cut at grouping things by related functionality, but this will undoubtedly keep changing.